src/librustc/ty/context.rs

```diff
@@ -2402,6 +2402,24 @@ impl<'tcx> TyCtxt<'tcx> {
         self.mk_generic_adt(def_id, ty)
     }
 
+    #[inline]
+    pub fn mk_pin(self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        let def_id = self.require_lang_item(lang_items::PinTypeLangItem, None);
```

Comment:
There's quite a bit of repetition here; You could introduce:
```rust
pub fn mk_lang_item(self, ty: Ty<'tcx>, item: LangItem)  -> Ty<'tcx> {
    let def_id = self.require_lang_item(item, None);
    self.mk_generic_adt(def_id, ty);
}
```
and then use this instead in the diagnostics logic.